---
title: "CCBR1046"
output:   
   flexdashboard::flex_dashboard:
   orientation: rows
   source_code: embed

---

```{r setup, include=FALSE}
library(tidyverse)
library(crosstalk)
library(DT)
```


Summary 
=======================================================================
### Annotatation Approach
CRISPResso2 output (Alleles_frequency_table.zip for SS1 as test) was read and aligned to the BRCA2 gene using pairwiseAlignment function with a “global-local” option.

Mismatch between Reference and input was considered either a mismatch or indel.

Variants were loaded into ANNOVAR and annotated for their SIFT, Polyphen, CADD, and CLINVAR Clinical Significance scores.

Variant Allele Frequnecy Table is the original output from CRISPResso2.  
Each row is linked its annotated ANNOVAR output in the second tab.

Variant Allele Frequency (CRISPResso)
=======================================================================

```{r}
vt_full=readRDS("vt_full.rds")
afch=readRDS("afch.rds")

shared_dt1=SharedData$new(afch[[1]],~af.id,group="Variant")
shared_dt2=SharedData$new(vt_full[[1]],~af.id,group="Variant")

#filter_select("af.id", "Variant:", shared_dt1, ~af.id)
```
### CRISPResso2 Allele Frequency Output

```{r}
DT::datatable(shared_dt1,rownames=F,
              extensions="Buttons",
              options=list( dom = 'Bfrtip')) %>%
  formatRound('Reads_prop', 5) 
```

ANNOVAR Variant Table
=======================================================================

### Annotated

```{r}
DT::datatable(shared_dt2,rownames=F,
              extensions="Buttons",
              options=list( dom = 'Bfrtip'))
```

